src/Microsoft.AspNetCore.Razor.GenerateTool/RunCommand.cs

```diff
@@ -0,0 +1,198 @@
+        {
+            var outputs = new OutputItem[sources.Count];
+            Parallel.For(0, outputs.Length, new ParallelOptions() { MaxDegreeOfParallelism = 4 }, i =>
+            {
+                var source = sources[i];
+                OutputItem compilationInfo;
+                using (var fileStream = source.CreateReadStream())
+                {
+                    var csharpDocument = templateEngine.GenerateCode(source.ViewEnginePath);
+                    compilationInfo = new OutputItem(source, csharpDocument);
+                }
+
+                outputs[i] = compilationInfo;
+            });
+
+            return outputs;
+        }
+
+        private bool ValidateArguments(Application application)
+        {
+            if (string.IsNullOrEmpty(application.OutputDirectory.Value()))
+            {
+                application.Error.WriteLine($"{application.OutputDirectory.ValueName} not specified.");
+                return false;
+            }
+
+            if (application.Sources.Values.Count == 0)
+            {
+                application.Error.WriteLine($"{application.Sources.Name} should have at least one value.");
+                return false;
+            }
+
+            return true;
+        }
+
+        private struct OutputItem
+        {
+            public OutputItem(
+                SourceItem viewFileInfo,
+                RazorCSharpDocument cSharpDocument)
+            {
+                ViewFileInfo = viewFileInfo;
+                CSharpDocument = cSharpDocument;
+            }
+
+            public SourceItem ViewFileInfo { get; }
+
+            public RazorCSharpDocument CSharpDocument { get; }
+        }
+
+        private struct SourceItem
+        {
+            public SourceItem(string fullPath, string viewEnginePath)
+            {
+                FullPath = fullPath;
+                ViewEnginePath = viewEnginePath;
+            }
+
+            public string FullPath { get; }
+
+            public string ViewEnginePath { get; }
+
+            public Stream CreateReadStream()
+            {
+                // We are setting buffer size to 1 to prevent FileStream from allocating it's internal buffer
+                // 0 causes constructor to throw
+                var bufferSize = 1;
+                return new FileStream(
+                    FullPath,
+                    FileMode.Open,
+                    FileAccess.Read,
+                    FileShare.ReadWrite,
+                    bufferSize,
+                    FileOptions.Asynchronous | FileOptions.SequentialScan);
+            }
+        }
+
+        private class StaticTagHelperFeature : ITagHelperFeature
+        {
+            public RazorEngine Engine { get; set; }
+
+            public IReadOnlyList<TagHelperDescriptor> TagHelpers { get; set; }
+
+            public IReadOnlyList<TagHelperDescriptor> GetDescriptors() => TagHelpers;
+        }
```

Comment:
This file basically contains what precompilation does today, but without emitting the assembly. Most of the logic for managing the file paths will need to move into MSBuild so we know which files to update/generate in a single place but this isn't a blocker.